Framework/CurveFitting/src/EigenComplexMatrix.cpp

```diff
@@ -0,0 +1,329 @@
+
+/// Set this matrix to identity matrix
+void ComplexMatrix::identity() { m_matrix.setIdentity(); }
+
+/// Set all elements to zero
+void ComplexMatrix::zero() { m_matrix.setZero(); }
+
+/// Set the matrix to be diagonal.
+/// @param d :: Values on the diagonal.
+void ComplexMatrix::diag(const ComplexVector &d) {
+  const auto n = d.size();
+  resize(n, n);
+  zero();
+  for (size_t i = 0; i < n; ++i) {
+    set(i, i, d.get(i));
+  }
+}
+
+/// add a matrix to this
+/// @param M :: A matrix
+ComplexMatrix &ComplexMatrix::operator+=(const ComplexMatrix &M) {
+  m_matrix += M.eigen();
+  return *this;
+}
+/// add a constant to this matrix
+/// @param d :: A number
+ComplexMatrix &ComplexMatrix::operator+=(const ComplexType &d) {
+  m_matrix.array() += d;
+  return *this;
+}
+/// subtract a matrix from this
+/// @param M :: A matrix
+ComplexMatrix &ComplexMatrix::operator-=(const ComplexMatrix &M) {
+  m_matrix -= M.eigen();
+  return *this;
+}
+/// multiply this matrix by a number
+/// @param d :: A number
+ComplexMatrix &ComplexMatrix::operator*=(const ComplexType &d) {
+  m_matrix *= d;
+  return *this;
+}
+
+/// Multiply this matrix by a matrix
+ComplexMatrix ComplexMatrix::operator*(const EigenMatrix &m) const {
+  if (m.size1() != size2()) {
+    throw std::invalid_argument("Matrix by matrix multiplication: matricies are of incompatible sizes.");
+  }
+
+  ComplexMatrix res(m.size1(), size2());
+  res.eigen() = eigen() * m.inspector();
+  return res;
+}
+
+/// Multiply this matrix by a complex matrix
+ComplexMatrix ComplexMatrix::operator*(const ComplexMatrix &m) const {
+  if (m.size1() != size2()) {
+    throw std::invalid_argument("Matrix by matrix multiplication: matricies are of incompatible sizes.");
+  }
+
+  ComplexMatrix res(m.size1(), size2());
+  res.eigen() = eigen() * m.eigen();
+  return res;
+}
+
+/// Solve system of linear equations M*x == rhs, M is this matrix
+/// This matrix is destroyed.
+/// @param rhs :: The right-hand-side vector
+/// @param x :: The solution vector
+void ComplexMatrix::solve(const ComplexVector &rhs, ComplexVector &x) {
+  if (size1() != size2()) {
+    throw std::invalid_argument("System of linear equations: the matrix must be square.");
+  }
+  size_t n = size1();
+  if (rhs.size() != n) {
+    throw std::invalid_argument("System of linear equations: right-hand side vector has wrong size.");
+  }
+  if (det() == ComplexType(0, 0)) {
+    throw std::invalid_argument("Matrix A is singular.");
+  }
+
+  Eigen::MatrixXcd b = rhs.eigen();
+  Eigen::ColPivHouseholderQR<Eigen::MatrixXcd> dec(eigen());
+  Eigen::VectorXcd res = dec.solve(b);
+  x = res;
+
+  if (!rhs.eigen().isApprox(eigen() * x.eigen())) {
+    throw std::runtime_error("Matrix Solution Error: solution does not exist.");
+  }
+}
+
+/// Invert this matrix
+void ComplexMatrix::invert() {
+  if (size1() != size2()) {
+    throw std::runtime_error("Matrix inverse: the matrix must be square.");
+  }
+  m_matrix = m_matrix.inverse();
```

Comment:
I'm having some issues with this. 

In previous implementation GSL would try to calculate the inverse regardless or not it exists. In doing so, it would mutate the input matrix.

If the linear decomposition exists, but the inverse doesn't, GSL would return the linear decomp matrix.

if the linear decomposition and the inverse doesn't exist, GSL would return the original matrix.

This behaviour is probably a bug, and not replicated here. I "fixed" the bug on another branch, but it appears atkease some tests rely on it